bundles/org.openhab.binding.enphase/src/main/java/org/openhab/binding/enphase/internal/handler/EnvoyBridgeHandler.java

```diff
@@ -0,0 +1,409 @@
+    private final EnvoyConnector connector;
+    private final EnvoyHostAddressCache envoyHostnameCache;
+
+    private @NonNullByDefault({}) EnvoyConfiguration configuration;
+    private @Nullable ScheduledFuture<?> updataDataFuture;
+    private @Nullable ScheduledFuture<?> updateHostnameFuture;
+    private @Nullable ExpiringCache<Map<String, @Nullable InverterDTO>> invertersCache;
+    private @Nullable ExpiringCache<Map<String, @Nullable DeviceDTO>> devicesCache;
+    private @Nullable EnvoyEnergyDTO productionDTO;
+    private @Nullable EnvoyEnergyDTO consumptionDTO;
+    private FeatureStatus consumptionSupported = FeatureStatus.UNKNOWN;
+    private FeatureStatus jsonSupported = FeatureStatus.UNKNOWN;
+
+    public EnvoyBridgeHandler(final Bridge thing, final HttpClient httpClient,
+            final EnvoyHostAddressCache envoyHostAddressCache) {
+        super(thing);
+        connector = new EnvoyConnector(httpClient);
+        this.envoyHostnameCache = envoyHostAddressCache;
+    }
+
+    @Override
+    public void handleCommand(final ChannelUID channelUID, final Command command) {
+        if (command instanceof RefreshType) {
+            refresh(channelUID);
+        }
+    }
+
+    private void refresh(final ChannelUID channelUID) {
+        final EnvoyEnergyDTO data = ENVOY_CHANNELGROUP_CONSUMPTION.equals(channelUID.getGroupId()) ? consumptionDTO
+                : productionDTO;
+
+        if (data == null) {
+            updateState(channelUID, UnDefType.UNDEF);
+        } else {
+            switch (channelUID.getIdWithoutGroup()) {
+                case ENVOY_WATT_HOURS_TODAY:
+                    updateState(channelUID, new QuantityType<>(data.wattHoursToday, Units.WATT_HOUR));
+                    break;
+                case ENVOY_WATT_HOURS_SEVEN_DAYS:
+                    updateState(channelUID, new QuantityType<>(data.wattHoursSevenDays, Units.WATT_HOUR));
+                    break;
+                case ENVOY_WATT_HOURS_LIFETIME:
+                    updateState(channelUID, new QuantityType<>(data.wattHoursLifetime, Units.WATT_HOUR));
+                    break;
+                case ENVOY_WATTS_NOW:
+                    updateState(channelUID, new QuantityType<>(data.wattsNow, Units.WATT));
+                    break;
+            }
+        }
+    }
+
+    @Override
+    public Collection<Class<? extends ThingHandlerService>> getServices() {
+        return Collections.singleton(EnphaseDevicesDiscoveryService.class);
+    }
+
+    @Override
+    public void initialize() {
+        configuration = getConfigAs(EnvoyConfiguration.class);
+        if (!EnphaseBindingConstants.isValidSerial(configuration.serialNumber)) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Serial Number is not valid");
+        }
+        if (configuration.hostname.isEmpty()) {
+            updateStatus(ThingStatus.ONLINE, ThingStatusDetail.CONFIGURATION_PENDING,
```

Comment:
Can it really be considered online at this point?

Reply:
Yes good question. I didn't know how to handle it otherwise. Looking at the status table: https://www.openhab.org/docs/concepts/things.html#status-details This comes closest to what would give some meaningful status in this case. Other status are not expected to have details and thus are not expected to have additional detail text. If you have another idea let me know.

Reply:
I've changed this so set to `UNKNOWN` at this point, and to set to OFF_LINE if it doesn't have an ip address.